### DISCUSSION TITLE
[TwigComponent] Parser fails on Twig tags within attributes

### DISCUSSION THREAD

**Discussion by ameotoko** [symfony/ux#2919] [Questions & Answers]
Normally, in a Twig template you can do something like this:

```twig
{% set target = 'target="_blank"' %}
<a
    href="{{ href|default('./') }}"
    title="{{ pageTitle|default(title) }}"
    {% if class|default %} class="{{ class }}"{% endif %}
    {% if tabindex|default %} tabindex="{{ tabindex }}"{% endif %}
    {{ target }}
>{{ link }}</a>
```

This doesn't seem possible with Twig Components, it breaks at `{% if` or `{{`, even if those are commented out:

```
Twig\Error\SyntaxError:
Expected attribute name when parsing the "<twig:Button" syntax.
```

It looks like the parser just bypasses the Twig engine completely and parses the code as raw file instead. I searched through the issues, and it seems you guys are already aware of the problem (for example here: https://github.com/symfony/ux/issues/880#issuecomment-1553275837) and it needs much more rewriting than one might think 🙂

Nevertheless, I thought that may be this specific use-case needs to be mentioned here, just to keep track of everything.

**Comment by WebMamba** [comment]
What are you trying to do? Something like:
```twig
<twig:Link {% if class|default %} class="{{ class }}"{% endif %} />
```
Or something similar ?

**Comment by ameotoko** [comment]
Yes, exactly that.

**Comment by smnandre** [comment]
You cannot, as you cannot either in include or embed component for instance. 

Because it would require a loop lexer / parser that does not exist.

In your example you "echo" something in Twig, but you cannot have logical "before" the execution of the component.

**Comment by WebMamba** [comment]
Hum yeah, I think we can't solve that easily! To have this TwigComponent syntax we have what we call a Prelexer. It's a little class that goes over your templates and transforms `<twig:` into `{{ component() }}`. The thing is in this Prelexer class we don't have access to all the tokens the Twig Lexer has access to. So for the moment, we let this syntax have it, but maybe one day the TwigComponent syntax will be fully integrated into Twig, but you need to convince Fabien first 😉

**Comment by Kocal** [comment]
That would be really hard to implement this syntax, but for the moment you can spread dynamic properties like that :
```twig
<twig:Link
    {{ ...class is defined ? { class } : {} }}
    {{ ...tabindex is defined ? { tabindex } : {} }}
/>

{# or #} 
<twig:Link {{ ...{
    class: class|default ? class : false,
    tabindex: tabindex|default ? tabindex : false,
} }}
/>
```

**Comment by smnandre** [comment]
@WebMamba honestly i'm not sure this is easily feasable (and personaly sure it should be done)

That would be equivalent to something like `{% if foo '{% if bar %}' %}  `

**Comment by smnandre** [comment]
> That would be really hard to implement this syntax, but for the moment you can spread dynamic properties like that :

What about 

```
<twig:Link class="{{ class ?? false }}" tabindex="{{ tabindex ?? false }}" /> 
```

**Comment by smnandre** [comment]
I'm closing here as this limitation is not something we really can work on

**Comment by ameotoko** [comment]
> What about
> 
> ```
> <twig:Link class="{{ class ?? false }}" tabindex="{{ tabindex ?? false }}" /> 
> ```

This actually does the right thing! It seems that I missed the part from the docs where it says: 

> Set an attribute's value to false to exclude the attribute:

```twig
{# templates/components/Input.html.twig #}
<input{{ attributes }}/>

{# render component #}
<twig:Input type="text" value="" :autofocus="false" />

{# renders as: #}
<input type="text" value=""/>
```
